staging/src/k8s.io/client-go/tools/events/event_broadcaster.go

```diff
@@ -0,0 +1,294 @@
+		sink:          sink,
+		randGen:       rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) refreshExistingEventSeries() {
+	eventBroadcaster.mu.RLock()
+	defer eventBroadcaster.mu.RUnlock()
+	for _, event := range eventBroadcaster.eventCache {
+
+		if event.Series != nil {
+			recordEvent(eventBroadcaster.sink, event, true)
+		}
+
+	}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) finishSeries() {
+	eventBroadcaster.mu.Lock()
+	defer eventBroadcaster.mu.Unlock()
+	for isomorphicKey, event := range eventBroadcaster.eventCache {
+		eventSerie := event.Series
+		if eventSerie != nil {
+
+			if eventSerie.LastObservedTime.Time.Add(6 * time.Minute).Before(time.Now()) {
+
+				if recordEvent(eventBroadcaster.sink, event, true) {
+					delete(eventBroadcaster.eventCache, isomorphicKey)
+				}
+			}
+		}
+
+	}
+}
+
+// NewRecorder returns an EventRecorder that records events with the given event source.
+func (eventBroadcaster *eventBroadcasterImpl) NewRecorder(scheme *runtime.Scheme, reportingController string) EventRecorder {
+
+	hostname, _ := os.Hostname()
+	reportingInstance := reportingController + "-" + hostname
+
+	return &recorderImpl{scheme, reportingController, reportingInstance, eventBroadcaster.Broadcaster, clock.RealClock{}}
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) recordToSink(event *v1beta1.Event, clock clock.Clock) {
+	// Make a copy before modification, because there could be multiple listeners.
+	// Events are safe to copy like this.
+	eventCopy := *event
+	event = &eventCopy
+
+	go func() {
+		eventKey := getKey(event)
+		eventBroadcaster.mu.RLock()
+		isomorphicEvent, isIsomorphic := eventBroadcaster.eventCache[eventKey]
+		eventBroadcaster.mu.RUnlock()
+
+		if isIsomorphic {
+
+			if isomorphicEvent.Series != nil {
+				isomorphicEvent.Series.Count++
+				isomorphicEvent.EventTime = metav1.MicroTime{Time: clock.Now()}
+			} else {
+				isomorphicEvent.Series = &v1beta1.EventSeries{
+					Count:            1,
+					LastObservedTime: metav1.MicroTime{Time: clock.Now()},
+				}
+				eventBroadcaster.attemptRecording(isomorphicEvent)
+			}
+
+		} else {
+			eventBroadcaster.mu.Lock()
+			eventBroadcaster.eventCache[eventKey] = event
+			eventBroadcaster.mu.Unlock()
+			eventBroadcaster.attemptRecording(event)
+		}
+
+	}()
+
+}
+
+func (eventBroadcaster *eventBroadcasterImpl) attemptRecording(event *v1beta1.Event) {
+	tries := 0
+
+	for {
+		if recordEvent(eventBroadcaster.sink, event, event.Series != nil) {
+			break
+		}
+		tries++
+		if tries >= maxTriesPerEvent {
+			klog.Errorf("Unable to write event '%#v' (retry limit exceeded!)", event)
+			break
+		}
+		// Randomize sleep so that various clients won't all be
+		// synced up if the master goes down.
+		time.Sleep(time.Duration(float64(eventBroadcaster.sleepDuration) * eventBroadcaster.randGen.Float64()))
+	}
+
+}
+
+func recordEvent(sink EventSink, event *v1beta1.Event, isEventSerie bool) bool {
```

Comment:
Why do we need to pass isEventSerie as a parameter? It's computed from "event != nil" in all places...